python/understack-workflows/understack_workflows/bmc_credentials.py

```diff
@@ -0,0 +1,142 @@
+    _set_bmc_creds(ip_address, token, username, required_password)
+    logger.info("BMC password has been set.")
+
+    if _verify_auth(ip_address, username, required_password):
+        logger.info("Production BMC credentials are working on this BMC.")
+
+
+def _verify_auth(host: str, username: str = "root", password: str = "") -> str:
+    """Test whether provided credentials work against a secured API endpoint.
+
+    Returns Session authentication token on success
+    Returns None on authorisation failure
+    Raises an Exception for other kinds of errors (e.g. timeout, etc)
+    """
+    try:
+        response = requests.request(
+            method="POST",
+            url=f"https://{host}/redfish/v1/SessionService/Sessions",
+            verify=False,
+            timeout=30,
+            json={"UserName": username, "Password": password},
+        )
+        if response.status_code == 401:
+            return None
+        if response.status_code >= 400:
+            raise Exception(
+                f"BMC {host} password login failed: "
+                f" {response.status_code} {response.json()}"
+            )
+            return None
+
+        return response.headers["X-Auth-Token"]
+    except Exception as e:
+        raise e from None
+
+
+def _get_bmc_accounts(host: str, token: str, username: str) -> list[dict]:
+    """A vendor agnostic approach to crawling the API for BMC accounts."""
+    try:
+        # get account service
+        r = _redfish_request(host, "/redfish/v1", token)
+        account_service_uri = r["AccountService"]["@odata.id"]
+        logger.debug(f"account_service_url: {account_service_uri}")
+
+        # get account collection uri
+        r = _redfish_request(host, account_service_uri, token)
+        accounts_uri = r["Accounts"]["@odata.id"]
+        logger.debug(f"accounts_url: {accounts_uri}")
+
+        # get accounts
+        r = _redfish_request(host, accounts_uri, token)
+        accounts = r["Members"]
+        logger.debug(f"accounts: {accounts}")
+
+        return accounts
+    except Exception:
+        logger.exception("Can't fetch accounts from Redfish account service.")
+        raise
+
+
+def _set_bmc_creds(host: str, token: str, username: str, new_password: str):
+    """Change password for the account associated with username in question."""
+    accounts = _get_bmc_accounts(host, token, username)
+
+    matched_account = None
+    for account in accounts:
+        account_url = account["@odata.id"]
+
+        a = _redfish_request(host, account_url, token)
+        if username == a["UserName"]:
+            logger.debug(f"found account: {a}")
+            matched_account = a
+            break
+
+    if not matched_account:
+        raise Exception(f"Unable to find BMC account for {username}")
+
+    account_uri = matched_account["@odata.id"]
+
+    payload = {"Password": new_password}
+    _redfish_request(host, account_uri, token, "PATCH", payload)
+
+
+def _redfish_request(
```

Comment:
doesn't this duplicate the same function from the Bmc class?

Reply:
It does, but they do different things - this code operates on tokens where the BMC class just uses basic auth.

Reply:
https://docs.openstack.org/sushy/latest/reference/usage.html#specifying-an-authentication-type